File: frontend/src/services/partner/revenue.ts
import { http } from "../../lib/http";

export interface RevenueSummary {
  total_revenue_minor: number;
  tenant_settlement_minor: number;
  kyradi_commission_minor: number;
  transaction_count: number;
  date?: string;
}

export interface Settlement {
  id: string;
  tenant_id: string;
  payment_id: string;
  reservation_id: string;
  total_amount_minor: number;
  tenant_settlement_minor: number;
  kyradi_commission_minor: number;
  currency: string;
  status: string;
  settled_at: string | null;
  commission_rate: number;
  created_at: string;
}

export interface SettlementListResponse {
  items: Settlement[];
  total_count: number;
  total_income: number;
  total_commission: number;
  total_payout: number;
  currency: string;
}

export interface PaymentModeRevenue {
  mode: string;
  label: string;
  total_revenue_minor: number;
  tenant_settlement_minor: number;
  kyradi_commission_minor: number;
  transaction_count: number;
}

export interface DailyRevenueItem {
  date: string;
  total_revenue_minor: number;
  tenant_settlement_minor: number;
  kyradi_commission_minor: number;
  transaction_count: number;
}

export interface RevenueHistoryResponse {
  items: DailyRevenueItem[];
  total_revenue_minor: number;
  total_tenant_settlement_minor: number;
  total_kyradi_commission_minor: number;
  total_transaction_count: number;
  period_start: string;
  period_end: string;
}

const normalizeArray = <T>(data: unknown): T[] => {
  if (Array.isArray(data)) return data as T[];
  if (data && typeof data === "object" && Array.isArray((data as { items?: unknown }).items)) {
    return (data as { items: T[] }).items;
  }
  return [];
};

const normalizeHistoryResponse = (data: unknown): RevenueHistoryResponse => {
  const fallback: RevenueHistoryResponse = {
    items: [],
    total_revenue_minor: 0,
    total_tenant_settlement_minor: 0,
    total_kyradi_commission_minor: 0,
    total_transaction_count: 0,
    period_start: "",
    period_end: "",
  };

  if (!data || typeof data !== "object") return fallback;
  const payload = data as Partial<RevenueHistoryResponse> & { items?: unknown };

  return {
    items: normalizeArray<DailyRevenueItem>(payload.items),
    total_revenue_minor: payload.total_revenue_minor ?? 0,
    total_tenant_settlement_minor: payload.total_tenant_settlement_minor ?? 0,
    total_kyradi_commission_minor: payload.total_kyradi_commission_minor ?? 0,
    total_transaction_count: payload.total_transaction_count ?? 0,
    period_start: payload.period_start ?? "",
    period_end: payload.period_end ?? "",
  };
};

const normalizeSettlementsResponse = (data: unknown): SettlementListResponse => {
  const fallback: SettlementListResponse = {
    items: [],
    total_count: 0,
    total_income: 0,
    total_commission: 0,
    total_payout: 0,
    currency: "TRY",
  };

  if (!data || typeof data !== "object") return fallback;
  const payload = data as Partial<SettlementListResponse> & { items?: unknown };

  return {
    items: normalizeArray<Settlement>(payload.items),
    total_count: payload.total_count ?? normalizeArray<Settlement>(payload.items).length,
    total_income: payload.total_income ?? 0,
    total_commission: payload.total_commission ?? 0,
    total_payout: payload.total_payout ?? 0,
    currency: payload.currency ?? "TRY",
  };
};

export const revenueService = {
  async getSummary(dateFrom?: string, dateTo?: string): Promise<RevenueSummary> {
    const params: Record<string, string> = {};
    if (dateFrom) params.from = dateFrom;
    if (dateTo) params.to = dateTo;
    const response = await http.get<RevenueSummary>("/revenue/summary", { params });
    return response.data;
  },

  async getDaily(date?: string): Promise<RevenueSummary> {
    const params: Record<string, string> = {};
    if (date) params.date = date;
    const response = await http.get<RevenueSummary>("/revenue/daily", { params });
    return response.data;
  },

  async listSettlements(
    status?: string,
    dateFrom?: string,
    dateTo?: string,
    locationId?: string,
    storageId?: string,
    search?: string
  ): Promise<SettlementListResponse> {
    const params: Record<string, string> = {};
    if (status) params.status = status;
    if (dateFrom) params.from = dateFrom;
    if (dateTo) params.to = dateTo;
    if (locationId) params.location_id = locationId;
    if (storageId) params.storage_id = storageId;
    if (search) params.search = search;
    const response = await http.get<SettlementListResponse>("/revenue/settlements", { params });
    return normalizeSettlementsResponse(response.data);
  },

  // Legacy method for backward compatibility
  async listSettlementsLegacy(
    status?: string,
    dateFrom?: string,
    dateTo?: string
  ): Promise<Settlement[]> {
    const params: Record<string, string> = {};
    if (status) params.status = status;
    if (dateFrom) params.from = dateFrom;
    if (dateTo) params.to = dateTo;
    const response = await http.get<Settlement[]>("/revenue/settlements/legacy", { params });
    return response.data;
  },

  async getByPaymentMode(dateFrom?: string, dateTo?: string): Promise<PaymentModeRevenue[]> {
    const params: Record<string, string> = {};
    if (dateFrom) params.from = dateFrom;
    if (dateTo) params.to = dateTo;
    const response = await http.get<PaymentModeRevenue[]>("/revenue/by-payment-mode", { params });
    return normalizeArray<PaymentModeRevenue>(response.data);
  },

  async getHistory(
    dateFrom?: string, 
    dateTo?: string, 
    granularity: "daily" | "weekly" | "monthly" = "daily"
  ): Promise<RevenueHistoryResponse> {
    const params: Record<string, string> = { granularity };
    if (dateFrom) params.from = dateFrom;
    if (dateTo) params.to = dateTo;
    const response = await http.get<RevenueHistoryResponse>("/revenue/history", { params });
    return normalizeHistoryResponse(response.data);
  },
};

File: backend/app/api/routes/revenue.py
"""Revenue and settlement endpoints."""

import logging
from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from ...db.session import get_session
from ...dependencies import require_accounting
from ...models import Settlement, User, Payment, Reservation, Location
from ...schemas.revenue import SettlementRead, RevenueSummary
from ...services.revenue import get_daily_revenue, get_tenant_revenue_summary, get_revenue_by_payment_mode

router = APIRouter(prefix="/revenue", tags=["revenue"])
logger = logging.getLogger(__name__)


class SettlementListResponse(BaseModel):
    """Response with settlements and totals."""
    items: List[SettlementRead]
    total_count: int
    total_income: int  # Minor units
    total_commission: int  # Minor units
    total_payout: int  # Minor units
    currency: str = "TRY"


class PaymentModeRevenue(BaseModel):
    """Revenue breakdown by payment mode."""
    mode: str
    label: str
    total_revenue_minor: int
    tenant_settlement_minor: int
    kyradi_commission_minor: int
    transaction_count: int


@router.get("/summary", response_model=RevenueSummary)
async def get_revenue_summary(
    date_from: Optional[datetime] = Query(default=None, alias="from"),
    date_to: Optional[datetime] = Query(default=None, alias="to"),
    current_user: User = Depends(require_accounting),
    session: AsyncSession = Depends(get_session),
) -> RevenueSummary:
    """Get revenue summary for the tenant."""
    summary = await get_tenant_revenue_summary(
        session,
        tenant_id=current_user.tenant_id,
        date_from=date_from,
        date_to=date_to,
    )
    return RevenueSummary(**summary)


@router.get("/daily", response_model=RevenueSummary)
async def get_daily_revenue_endpoint(
    date: Optional[datetime] = Query(default=None),
    current_user: User = Depends(require_accounting),
    session: AsyncSession = Depends(get_session),
) -> RevenueSummary:
    """Get daily revenue for today or specified date."""
    daily = await get_daily_revenue(
        session,
        tenant_id=current_user.tenant_id,
        date=date,
    )
    return RevenueSummary(**daily)


@router.get("/by-payment-mode", response_model=List[PaymentModeRevenue])
async def get_revenue_by_payment_mode_endpoint(
    date_from: Optional[datetime] = Query(default=None, alias="from"),
    date_to: Optional[datetime] = Query(default=None, alias="to"),
    current_user: User = Depends(require_accounting),
    session: AsyncSession = Depends(get_session),
) -> List[PaymentModeRevenue]:
    """Get revenue breakdown by payment mode (POS, CASH, GATEWAY)."""
    data = await get_revenue_by_payment_mode(
        session,
        tenant_id=current_user.tenant_id,
        date_from=date_from,
        date_to=date_to,
    )
    return [PaymentModeRevenue(**item) for item in data]


@router.get("/settlements", response_model=SettlementListResponse)
async def list_settlements(
    status_filter: Optional[str] = Query(default=None, alias="status"),
    date_from: Optional[datetime] = Query(default=None, alias="from"),
    date_to: Optional[datetime] = Query(default=None, alias="to"),
    location_id: Optional[str] = Query(default=None),
    storage_id: Optional[str] = Query(default=None),
    search: Optional[str] = Query(default=None),
    current_user: User = Depends(require_accounting),
    session: AsyncSession = Depends(get_session),
) -> SettlementListResponse:
    """List settlements for the tenant with totals."""
    from ...models import Storage
    
    stmt = select(Settlement).where(Settlement.tenant_id == current_user.tenant_id)
    
    if status_filter:
        stmt = stmt.where(Settlement.status == status_filter)
    if date_from:
        stmt = stmt.where(Settlement.created_at >= date_from)
    if date_to:
        stmt = stmt.where(Settlement.created_at <= date_to)
    
    # Filter by location or storage via payment -> reservation -> storage -> location
    if location_id or storage_id:
        from sqlalchemy import and_
        # Build subquery with proper JOINs to avoid cartesian product
        # Use explicit select_from to make JOIN conditions clear
        payment_subquery = select(Payment.id)
        
        # Build the FROM clause explicitly with JOINs
        if location_id:
            # Need to join Storage to access location_id
            payment_subquery = payment_subquery.select_from(
                Payment.join(Reservation, Payment.reservation_id == Reservation.id)
                       .join(Storage, Reservation.storage_id == Storage.id)
            ).where(
                Payment.tenant_id == current_user.tenant_id,
                Reservation.tenant_id == current_user.tenant_id,
                Storage.location_id == location_id,
            )
        elif storage_id:
            # Only need Reservation for storage_id filter
            payment_subquery = payment_subquery.select_from(
                Payment.join(Reservation, Payment.reservation_id == Reservation.id)
            ).where(
                Payment.tenant_id == current_user.tenant_id,
                Reservation.tenant_id == current_user.tenant_id,
                Reservation.storage_id == storage_id,
            )
        else:
            # Just join Payment and Reservation
            payment_subquery = payment_subquery.select_from(
                Payment.join(Reservation, Payment.reservation_id == Reservation.id)
            ).where(
                Payment.tenant_id == current_user.tenant_id,
                Reservation.tenant_id == current_user.tenant_id,
            )
        
        stmt = stmt.where(Settlement.payment_id.in_(payment_subquery))
    
    # Search by payment_id or settlement id
    if search:
        stmt = stmt.where(
            Settlement.payment_id.ilike(f"%{search}%") |
            Settlement.id.ilike(f"%{search}%")
        )
    
    stmt = stmt.order_by(Settlement.created_at.desc())
    
    result = await session.execute(stmt)
    settlements = result.scalars().all()
    
    # Calculate totals
    total_income = sum(s.total_amount_minor for s in settlements)
    total_commission = sum(s.kyradi_commission_minor for s in settlements)
    total_payout = sum(s.tenant_settlement_minor for s in settlements)
    
    logger.info(
        f"Listed {len(settlements)} settlements for tenant {current_user.tenant_id}, "
        f"total_income={total_income}, total_commission={total_commission}, total_payout={total_payout}"
    )
    
    return SettlementListResponse(
        items=[SettlementRead.model_validate(s) for s in settlements],
        total_count=len(settlements),
        total_income=total_income,
        total_commission=total_commission,
        total_payout=total_payout,
        currency="TRY",
    )


@router.get("/settlements/legacy", response_model=List[SettlementRead])
async def list_settlements_legacy(
    status: Optional[str] = Query(default=None),
    date_from: Optional[datetime] = Query(default=None, alias="from"),
    date_to: Optional[datetime] = Query(default=None, alias="to"),
    current_user: User = Depends(require_accounting),
    session: AsyncSession = Depends(get_session),
) -> List[SettlementRead]:
    """List settlements for the tenant (legacy format without totals)."""
    stmt = select(Settlement).where(Settlement.tenant_id == current_user.tenant_id)
    
    if status:
        stmt = stmt.where(Settlement.status == status)
    if date_from:
        stmt = stmt.where(Settlement.settled_at >= date_from)
    if date_to:
        stmt = stmt.where(Settlement.settled_at <= date_to)
    
    stmt = stmt.order_by(Settlement.created_at.desc())
    
    result = await session.execute(stmt)
    settlements = result.scalars().all()
    
    return [SettlementRead.model_validate(settlement) for settlement in settlements]

